examples/with-apollo-hasura-ssr/lib/withAuth.tsx

```diff
@@ -0,0 +1,111 @@
+import React, { useEffect, useMemo } from 'react'
+import { UserProvider } from '../components/UserProvider'
+import auth0 from './auth0'
+import createLoginURL from './createLoginURL'
+import fetchUserAtClient from './fetchUserAtClient'
+
+type HasUser = Readonly<{ user: IClaims }>
+
+const loginPath = '/api/login'
+
+/**
+ * Only displays this page if the user is authenticated, otherwise redirects.
+ * @param PageComponent The page to wrap with Auth — will require users to login before seeing it.
+ */
+export default function withAuth<P extends Record<string, any> = {}, IP = P>(
+  PageComponent: NextPage<P, IP>,
+  log = false
+): NextPage<P, IP> {
+
+  // This is the component that renders in the browser or on the server, before
+  // PageComponent renders. So ensure that we don't render the PageComponent if
+  // the user is missing.
+  function WithAuth({ user, ...pageProps }: HasUser & P) {
+    const router = useRouter()
+    const value = useMemo(() => ({ user, loading: false }), [ user ])
+
+    // This runs client-side, and redirects the browser
+    useEffect(() => {
+      if (log) console.info('user == null; redirecting to login URL from WithAuth.useEffect')
+      if (typeof window === 'undefined') return
+      if (user == null) window.location.assign(createLoginURL(loginPath, router.asPath))
+    }, [ user, router ])
+
+    if (user == null) {
+      return 'Redirecting to sign-in page...'
+    }
+
+    if (log) console.info('WithAuth rendering UserProvider.Value=', value == null ? 'null' : 'not null')
+
+    return <UserProvider.Provider value={value}>
+      <PageComponent {...pageProps as any} />
+    </UserProvider.Provider>
+  }
+
+  // Be nice to developers:
+  if (process.env.NODE_ENV !== 'production') {
+    const displayName =
+      PageComponent.displayName || PageComponent.name || 'Component'
+
+    if (displayName === 'App') {
+      console.warn('This withAuth HOC only works with PageComponents.')
+    }
+
+    WithAuth.displayName = `withAuth(${displayName})`
+  }
+
+  async function getPageProps(ctx: NextPageContext) {
+    let pageProps: Record<string, any> = {}
+    if (PageComponent.getInitialProps != null) {
+      pageProps = await PageComponent.getInitialProps(ctx)
+    }
+    return pageProps
+  }
+
+  WithAuth.getInitialProps = async (ctx: NextPageContext) => {
+    if(log) console.info('WithAuth.getInitialProps initial')
+
+    const req = ctx.req, res = ctx.res // to help TypeScript
+
+    // Client-side rendering
+    if (req == null || res == null) {
+      if (log) console.info('WithAuth.getInitialProps client-side rendering case')
+      const pageProps = await getPageProps(ctx)
+      return {
+        ...pageProps,
+        user: 'user' in pageProps ? pageProps.user : await fetchUserAtClient(),
+      }
+    }
+
+    // @ts-ignore
+    const session = await auth0.getSession(req)
+
+    // Server-side, unauthenticated, just returns the location header
+    if (!session || !session.user) {
```

Comment:
This is also true, if the session is expired. Is this the expected behavior? Otherwise, I would propose to validate the expiration date. 

```suggestion
    const isExpired = Date.now() > session?.accessTokenExpiresAt * 1000 - 60000
    if (!session || !session.user || isExpired) {
```

Reply:
Maybe we should move `withAuth` into the library proper as well?

Reply:
Also, you can put this extra if inside the parent if, since you qualify `session` as non-null in the leftmost disjunctions.